### ISSUE TITLE
Dataflow: missing catch blocks when handling "non-throwing" instructions 

### ISSUE THREAD

**Issue by artempyanykh** [typetools/checker-framework#4018]
Example:
```java
try {
  return new int[n];
} catch (OutOfMemoryError e) {
  // This catch is missing in the CFG
  return null;
}
```

Generated CFG (*note absence of exceptional flow*):
```
2 -> 3 EACH_TO_EACH
3 -> 0 EACH_TO_EACH

2:
Process order: 1
<entry>

3:
Process order: 2
marker (start of try statement #0)   [ Marker ]
marker (start of try block #0)   [ Marker ]
n   [ LocalVariable ]
new int[] (n)   [ ArrayCreation ]
return new int[] (n)   [ Return ]

0:
Process order: 3
<exit>
```

This is something I'd ideally like to fix. But at the same time I understand this is a somewhat tricky case since object creation/allocation is assumed to be non-throwing. 

Not sure what would be the best way to proceed here. Would appreciate any thoughts on the subject!

**Comment by wmdietl** [comment]
@artempyanykh Thanks for raising this!

When building dataflow we had decided to not add all possible `Error` edges that could arise:
https://github.com/typetools/checker-framework/blob/master/dataflow/manual/content.tex#L853
Adding all these edges would likely cause many false positives.

However, if there is a catch block for an `Error` type, we should make sure it is reachable.
For the Checker Framework, this again can only lead to false positives, as illustrated by this example:

````java
import org.checkerframework.checker.nullness.qual.Nullable;
class Demo {
  void foo(@Nullable Object nbl) {
    try {
    } catch (OutOfMemoryError e) {
      if (nbl != null) {
        nbl.toString();
      }
    }
  }
}
````

Because the catch block isn't reachable in the CFG, we don't refine the type of `nbl` and raise a spurious error.

What do you think of the following change: at the beginning of a try block we add an edge to each catch block for `Error` types.
So the above example would look like:

````java
import org.checkerframework.checker.nullness.qual.Nullable;
class Demo {
  void foo(@Nullable Object nbl) {
    try {
      if (*) {
        throw new OutOfMemoryError();
      }
    } catch (OutOfMemoryError e) {
      if (nbl != null) {
        nbl.toString();
      }
    }
  }
}
````

For this we would need some non-deterministic conditional jump edge or a condition-throw construct as above.

Optionally, we could add edges from all the places where an `Error` could be thrown. As even the empty try block can cause an `Error`, this isn't enough, but it would additionally precisely handle all locations that can throw.
For our use in the Checker Framework, I don't think this would be useful.
Would you need that precision for your use of dataflow?

**Comment by mernst** [comment]
@artempyanykh Could you please take a look at #4022?  It might solve your problem.

**Comment by artempyanykh** [comment]
Thanks for your input @wmdietl and @mernst! 

To give you a bit more colour - the tool I'm working on offloads *all* type-inference to forward analysis over CFG. So when the CFG is incomplete I have trees without qualified types which is problematic. Fortunately, there's not that much code in our codebase for which CFG is partial, so I could work around the problem at a different level for now.

I agree that a general solution to the problem would probably be too heavy for the extra benefit. But `nodesWithException` actually looks like just what the doctor ordered.

@mernst #4022 is very helpful, thank you! In my case it's not just `OOMError`s that are problematic. But let me see if I can fit whatever else is there into `nodesWithException`.

**Comment by mernst** [comment]
@artempyanykh I am inclined to merge #4022 because it is an improvement that doesn't seem to have a downside.  But please let me know if you have any comments or suggestions regarding it.

**Comment by artempyanykh** [comment]
@mernst please do merge. Sorry if I wasn't clear enough - #4022 will fix a number of partial CFG cases for me, so this is an improvement. And then I can follow the same approach to fix whatever else of similar type I find in our codebase.

**Comment by mernst** [comment]
Thanks for the clarification.  Please let us know if you notice other issues that should be fixed in the Checker Framework.  We appreciate the bug reports and pull requests.

**Comment by wmdietl** [comment]
The false positive from https://github.com/typetools/checker-framework/issues/4018#issuecomment-744861768 isn't fixed by #4022 and there are many other "non-throwing" instructions not fixed by #4022.
Why close this issue before it is fixed?

**Comment by mernst** [comment]
#4022 fixes the test case given in the original post.  Your comment is a good one (it was written after the fix in #4022 was completed, but before the fix had passed CI).  I feel it should be opened in a separate issue.  Do you want to do that, or should I?

**Comment by wmdietl** [comment]
The issue title is general, and the issue isn't limited to the OOME used in the example.
So I don't know how a new issue would differ from this issue and would simply re-open this one.

**Comment by mernst** [comment]
The issue *content* is not general.  A new issue can be focused on just the part of the problem that remains.  It will be much easier to interpret, since people won't have to wade through a lot of unhelpful conversations (like this one :-) to understand the current status.
